### DISCUSSION TITLE
Sqlite map from mapc2mapc - not working

### DISCUSSION THREAD

**Discussion by mystichussar** [osmandapp/OsmAnd#11735] [Q&A]
Hi! I followed one of tutorial on yt of a guy Bart Eisenberg, where he was converting old maps to georeferenced ones. The result was a sqlite file, which supposed to be readable by OSMAnd, but unfortunately it is not. At least not form me.

I uploaded file, selected "Online maps", set my map as under/over-lay, but nothing is changing. 

I also check that sqlite file in Locus and it works right away.

What am I doing wrong? Is it a bug?

Cheers

**Comment by barteis** [comment]
I'm seeing the same thing.  It still works with a single  level of tiles at a low zoom, but ly seemingnot multiple levels or smaller, more detailed tiles. Perhaps the file size is a problem.  I made the video four years ago, demonstrating how to create OsmAnd-compatible sqlite files using MAPC2MAPC. It of course worked fine back then. 

Here's the video: https://youtu.be/Y_fekLfcUOc

**Reply by barteis** [reply]
@mystichussar et.al.  Further experimenting:  I'm seeing the SQLite image on OsmAnd as long as I'm sufficiently zoomed in.  But I'm having trouble getting an exact calibration.  I'm wondering if the problem might  be with the current version of MAPC2MAPC.

**Reply by mariush444** [reply]
@mystichussar Could you attache/share example of map which you want to convert?
I can try to convert in my way then you can compare results and distinguish if the problem is with m2m or osmand.

**Reply by barteis** [reply]
![Bolinas](https://user-images.githubusercontent.com/36173950/120032668-62a38e00-bfaf-11eb-9a1d-c1fe6bf04458.jpg)

Here's an example--an historic topo map of and around Bolinas, California.  Thanks!

**Reply by mariush444** [reply]
Do you have georeferences to this file? There are no lat/lon on the picture.

**Reply by barteis** [reply]
If I georeference it, I'd ordinarily be using MAPC2MAPC, which I'm having some trouble with.  Should I try it in  QGIS?  If so, what format should the output file be?  Thanks.

**Reply by mariush444** [reply]
So probably the problem is to guess right EPSG for this map. Do you now when (the year) the map was published?
It could be useful to exclude some EPSG from the potential list.
QGIS most probably has more possibilities than m2m but I don't know m2m to much.
You should use the option (in the screenshot) for creating tiles in QGIS
![image](https://user-images.githubusercontent.com/66887280/120049240-a92dd280-c019-11eb-8378-8ca083d20e2a.png)

**Reply by mystichussar** [reply]
@mariush444 let try with that one:
http://maps.mapywig.org/m/German_maps/series/025K_TK25/2577_Starowitz_geognostisch-agronomische_1909.jpg

**Reply by barteis** [reply]
The map was published in 1993 and I'm pretty sure I found the right EPSG--QGIS has some helpful guidance for NAD27.  But I'm out of practice using the QGIS georeferencer.  I'm now practicing.

**Reply by barteis** [reply]
@mariush444 I'm not sure this is what you're looking for, but I produced this GeoTiff in QGIS of the above map.  I tried converting it to a sqlite file via MAPC2MAPC, but OsmAnd didn't show it.  Github doesn't accept tif files, but here's a link:
https://drive.google.com/file/d/1l06UMQppRhgF9PI1bv29DiaFO8ujG8Qp/view?usp=sharing

**Reply by barteis** [reply]
Sorry--I thought that was the default. I changed the permission to anyone with the link.

**Reply by mariush444** [reply]
so, finally
sqlitedb
https://drive.google.com/file/d/1aMN2XuykUfifUdb5UiRiFvYPI0tVJ_dl/view?usp=sharing
directory of tiles
https://drive.google.com/file/d/186Krkz1oX8jNTTrMigWz_DDaqDlWf3VQ/view?usp=sharing

**Reply by barteis** [reply]
That works on OsmAnd!  How did you produce it?  What's the proper way to set the CRS in QGIS?  Thanks,

**Reply by mariush444** [reply]
I started OSM layer with EPSG:3857 then I added raster layer with your tif. QGIS sees this as CRS NAD27.
It seems to be correct but map is somewhere at Hawaii. So in properties of the tif layer I changed CRS/Source to EPSG:3857
Then I can see your map. I tried to export it directly to tiles but it was still wrong. So I exported this tif layer (right click on the layer) ... saved as geotiff with CRS 3857. 
Then 
`gdal2tiles.py --xyz -x new-tif.tif dest-dir
`

**Reply by mystichussar** [reply]
@barteis @mariush444 So what's the conclusion:
a) is the OSMAnd OK and mapc2mapc sucks;
b) mapc2mapc is OK and OSMAnd sucks;
c) both of these are messing up giving us flaky results

How can we manage it to be doable for a newbie like me? I do have Linux instance on some of my PC and I think I could try mariush444's instruction. Or perhaps it doesn't have to be Linux OS?

**Reply by barteis** [reply]
@mariush444 I went back to the QGIS georeferencer and started over, this time making sure that both the OSM layer _and_ my map were set to EPSG:3857.  I generated a tiff file.  And then I used MAPC2MAPC to generate .sqlitedb file.  It worked: displaying perfectly on OsmAnd.  
https://drive.google.com/file/d/1B6TJovp2bw6KnlnPLMk5wzmniRvv_mRX/view?usp=sharing
 
@mystichussar 
At least in this case, MAPC was able to translate a sqlite file georeferenced in QGIS so that it is displayable on OsmAnd.  But I was not able to do the same using MAPC's georeferencing capability.  That could be me, or could be MAPC--I'm not sure.  If I get a chance, I'll write to John Thorn, MAPC's author, asking for his take.  

Your point about making this doable is a good one.  For me, the advantage of MAPC is that it's an all-in-one solution.  But, it's only for Windows and it's not fee, though I think Thorn charges a reasonable price for it.  

For the while, my immediate interest is in whether the method outlined in my [video tutorial](https://youtu.be/Y_fekLfcUOc) from four years ago still works.

**Reply by mariush444** [reply]
I'm user (not specialist) as other user so it is my opinion/observations only.
There is nothing about m2m or Osmand.
There is something with default settings of dgal (DSoG).

1. m2m is GUI for gdal (AFAIK) and uses DSoG
2. QGIS (in standard way) uses DSoG
3. I (my scripts) use DSoG
All 3 methods create the same map somewhere on the north to the Hawai.
It is done because inside the geotif there is information 

```
PROJCRS["NAD27 / California zone III",
    BASEGEOGCRS["NAD27",
        DATUM["North American Datum 1927",
            ELLIPSOID["Clarke 1866",6378206.4,294.978698213898,
                LENGTHUNIT["metre",1]]],
```
So gdal in these 3 solution does what it should do.

What I did (as I understand) I said inside QGIS "don't use NAD27, try to use WGS84"
QGIS did it and during this step the a conversasion was done (I guess).
The new-tif created with this step could be use by m2m,QGIS,my-scripts with DSoG to produce map in right place.
You can get the new-tif and generate new map with m2m without problem.

Is there a general way to make solution like simple manual for every case? I'm not sure.

**Reply by barteis** [reply]
I'm pretty sure now that my problem with MAPC is related to running a newer version  it on my particular hardware and setup.  When I use version 573, which I've used  for years, it seems to work fine--producing sqlite files that work as expected on OsmAnd.  John Thorn requested more info in order to troubleshoot.

**Reply by mystichussar** [reply]
Hi @barteis , there is something wrong with yt commenting section, and I couldn't post a comment (comments section refreshes in a loop?) so I post my sqlite here: https://we.tl/t-qQGboJBqVO

**Reply by barteis** [reply]
@mystichussar Thanks.  I got it to my phone, but can't figure out where in the world to look for it.  Where is it?

**Reply by mystichussar** [reply]
Uhm yes, sorry. It's 53'30"N 18'20"E

**Reply by barteis** [reply]
@mystichussar Unfortunately, I'm not seeing the map on OsmAnd either.  I'm still working with Thorn to resolve my issue with MAPC.  Maybe that will shed some light.

**Comment by mariush444** [comment]
If you can install gdal on your PC then you can convert your maps by yourself.
But the conversion is to tiles in directory and not to SQLite.
Just 3 commands
```
        gdalwarp -overwrite  -of GTiff your_map tmp.tif
        gdal_translate -of vrt -expand rgba tmp.tif tmp.vrt
        gdal2tiles.py --xyz -x tmp.vrt directory_for_tiles_made_from_your_map
```

**Reply by barteis** [reply]
Interesting.  What's the input file?  Is it a geotiff?  And how do you produce that?  Does QGIS do it?

**Reply by thompehw** [reply]
I think QGIS can do the equivalent generation of a tile hierarchy after any earlier steps required. It doesn't need the QTILES plugin any more, just do Processing>Toolbox>Raster tools>Generate XYZ tiles (Directory).

But, annoyingly, for OsmAnd you need to recurse through the resulting directory hierarchy and rename .jpg->.jpg.tile or .png->.png.tile

@mariush444 you forgot to mention this renaming step, which your own bash script that you were kind enough to share recently in #9944 does as a final step. Your script shows how to do it in Linux. 

In Windows, the recursive file renaming is awkward but I found a relatively easy way. If anyone's interested I can share it.

**Reply by mariush444** [reply]
@thompehw , yes you are right about the last step "renaming". I've forgot to mention this.
@barteis , so, I use linux. When you install QGIS then you have ( gdal https://gdal.org ) with this 3 needed commands.
It means that you can use QGIS (GUI) or gdal commands to convert your maps into tiles in directory.
Your maps can be GeoTiff, GeoPdf, Ozi (map/ozf2/gif/jpg).

If you have Linux then you can try to use my script 
https://github.com/mariush444/raster2osmand

**Reply by vshcherb** [reply]
I didn't understand why it needs to be renamed... You can simply change url patter or .metainfo file and that's it https://docs.osmand.net/en/main@latest/development/osmand-file-formats/osmand-metainfo

I see.. https://github.com/osmandapp/Osmand/blob/master/OsmAnd-java/src/main/java/net/osmand/map/TileSourceManager.java#L406 - most of OsmAnd version store with ".tile" in the end. So it doesn't matter if upgrade now cause it needs to be backward compatible anyway.

**Reply by thompehw** [reply]
@vshcherb  would it be possible to recognise tiles both with and without .tile at the end? Backwards compatibility, and better compatibility with other tile generators.

It would only benefit some advanced users, but it would make the flow from QGIS easy.

**Reply by mariush444** [reply]
@thompehw  @vshcherb This is good idea. I find only one pros for .tile for this moment 
Thanks to it  the files png are not indexed as all others jpg/png/gif on Android as pcitures/photo.
It would be very heavy process for thousands of pngs.
But there is solution, just putting .nomedia file to the directory. (into main OSMAND directory should be enough)

**Reply by vshcherb** [reply]
Well it's possible to have solution in future version (we will need request for that) that will be compatible with old maps, but it's impossible to that new maps will work in old versions, so it becomes quite tricky but anyway that will take time to properly implement.

**Reply by karl-01** [reply]
I have a side-question to this topic:
You generate Sqlite-maps or such maps with tiles on a PC.
But how do you copy these data to the Osmand-directory under android 11?
On Android 10 it is no problem, as you can use a filemanager.
But what with android 11??

Thanks

**Reply by mariush444** [reply]
I havn't got Android 11 but probably you should have possibility to set "data storage folder" in Osmand menu to multiuser storage 1.
Obb directory should be accessible. (I don't know, I just guess, sorry). I use TotalCommander as a file manager.

**Reply by mystichussar** [reply]
@karl-01 I have android 10 onboard

**Reply by barteis** [reply]
Any answers to the original topic of this thread?  The current version of OsmAnd is having trouble displaying multi-zoom level sqlite files.  Or perhaps any sqlite file beyond a certain size.   This once seemed to work flawlessly.

**Reply by mariush444** [reply]
@barteis  can you attach any example? Maybe the problem is with m2m and not with osmand.
Did you try to do it directly via gdal to compare results?

**Generally I think that the main problem is that Osmand doesn't support GeoTiff and/or GeoPdf.
In my opinion it is not customer oriented solution that such basic need is not supported.
Generating tiles/mbtiles is workaround only. Then you have 300MB of thousends tiles insted of 4MB ONE file.
It is not user friendly that user have to implement other apps/solutions/workaround just to see one file in osmand.**

So, please vote for https://github.com/osmandapp/OsmAnd/issues/9944

**Reply by vshcherb** [reply]
@barteis The current version of OsmAnd is having trouble displaying multi-zoom level sqlite files.  Not sure if I could catch what do you mean. 

Obviously GeoPdf, GeoTiff and other tiff images could be probably builtin as single utility in OsmAndMapCreator first. It's quite unlikely that OsmAnd will change format for Online (sqlite) & Vector (obf) maps cause that would require quite a lot of work.

**Reply by mariush444** [reply]
@vshcherb It is not about to change format for online maps but it is about to add possibility to use offline GeoPdf, GeoTiff.
That is all.
Why user have to use any conversion mapc2mapc/gdal/OsmAndMapCreator just to see calibrated maps?
Why user have to use bigger SD card for so huge (number and size) tiles instead to use small SD for small GeoFile?
Why user have to pay for mapc2mapc or develop scripts just to see simple map? Is Osmand for users or developers?
Don't get me wrong, please. It is of course just my opinion / point of view.
There would be no question about converting old maps via mapc2mapc and problems with tiles if GeoFile was supported.

**Reply by thompehw** [reply]
Sorry for continuing the drift off topic, but:

@mariush444 have you experimented with ways of making the tiles more compact? In my experience with historical maps, if you use jpg tiles instead of png you can use a high compression ratio without it making a visible difference. And I take care not to generate tiles for more zoom levels than I need. OsmAnd doesn't display tiles at a lower zoom level than they are provided, but it does continue to display enlarged images of the highest zoom tiles as you zoom beyond what you have generated. For an old map of a town, I found generating only levels 14-17 was adequate for me (using it as an underlay in OsmAnd).

I hope that's not too obvious! Best wishes.
